examples/seq2seq/utils.py

```diff
@@ -99,35 +116,47 @@ def __init__(
         tok_name = tokenizer.__class__.__name__.lower().rstrip("tokenizer")
         if hasattr(tokenizer, "set_lang") and src_lang is not None:
             tokenizer.set_lang(src_lang)  # HACK: only applies to mbart
-        self.source = encode_file(
-            tokenizer,
-            os.path.join(data_dir, type_path + ".source"),
-            max_source_length,
-            overwrite_cache=overwrite_cache,
-            prefix=prefix,
-            tok_name=tok_name,
-        )
-        tgt_path = os.path.join(data_dir, type_path + ".target")
-        if hasattr(tokenizer, "set_lang"):
+        self.max_source_length = max_source_length
+        self.max_target_length = max_target_length
+        self.len = self._get_examples(os.path.join(data_dir, type_path + ".source"))
+        self.source_file = os.path.join(data_dir, type_path + ".source")
+        self.tgt_file = os.path.join(data_dir, type_path + ".target")
+        self.tokenizer = tokenizer
+        
+        if hasattr(self.tokenizer, "set_lang"):
             assert tgt_lang is not None, "--tgt_lang must be passed to build a translation"
-            tokenizer.set_lang(tgt_lang)  # HACK: only applies to mbart
-        self.target = encode_file(
-            tokenizer, tgt_path, max_target_length, overwrite_cache=overwrite_cache, tok_name=tok_name
-        )
+            self.tokenizer.set_lang(tgt_lang)  # HACK: only applies to mbart
+        
         if n_obs is not None:
-            self.source = self.source[:n_obs]
-            self.target = self.target[:n_obs]
-        self.pad_token_id = tokenizer.pad_token_id
+            self.len = n_obs
+        self.pad_token_id = self.tokenizer.pad_token_id
 
     def __len__(self):
-        return len(self.source)
+        return self.len
 
     def __getitem__(self, index):
-        source_ids = self.source[index]["input_ids"].squeeze()
-        target_ids = self.target[index]["input_ids"].squeeze()
-        src_mask = self.source[index]["attention_mask"].squeeze()
+        source_line = linecache.getline(self.source_file, index).rstrip("\n")
+        tgt_line = linecache.getline(self.tgt_file, index).rstrip("\n")
+        source_inputs = encode_line(self.tokenizer,
+            source_line,
+            self.max_source_length)
+        
+        target_inputs = encode_line(self.tokenizer,
+            tgt_line,
+            self.max_target_length)
+
+        source_ids = source_inputs['input_ids'].squeeze()
+        target_ids = target_inputs['input_ids'].squeeze()
+        src_mask =source_inputs["attention_mask"].squeeze()
         return {"input_ids": source_ids, "attention_mask": src_mask, "decoder_input_ids": target_ids}
 
+    @staticmethod
+    def _get_examples(data_file):
```

Comment:
Yeah I considered that one too. But since I also want to get raw sequence lengths for the sortish sampler anyway, my `_get_examples()` method does both simultaneously. See diffs on the latest commits.